### PR TITLE
Default sync

### PR DESCRIPTION
**Pull request by bryevdv**
Replaces/finishes #3414 is based on #3416 so shows those commits as well.

ping @mattpap @havocp 

Tested with all server apps and examples. 

Current report:
```
BoxSelectTool: coffee is missing some properties found in Python
    BoxSelectTool.overlay: python defaults to {"attributes":{"bottom":null,"bottom_units":"screen","fill_alpha":{"value":0.5},"fill_color":{"value":"lightgrey"},"left":null,"left_units":"screen","level":"overlay","line_alpha":{"value":1},"line_cap":"butt","line_color":{"value":"black"},"line_dash":[4,4],"line_dash_offset":0,"line_join":"miter","line_width":{"value":2},"name":null,"plot":null,"render_mode":"css","right":null,"right_units":"screen","tags":[],"top":null,"top_units":"screen","x_range_name":"default","y_range_name":"default"},"type":"BoxAnnotation"} but coffee has no such property
BoxZoomTool: coffee is missing some properties found in Python
    BoxZoomTool.overlay: python defaults to {"attributes":{"bottom":null,"bottom_units":"screen","fill_alpha":{"value":0.5},"fill_color":{"value":"lightgrey"},"left":null,"left_units":"screen","level":"overlay","line_alpha":{"value":1},"line_cap":"butt","line_color":{"value":"black"},"line_dash":[4,4],"line_dash_offset":0,"line_join":"miter","line_width":{"value":2},"name":null,"plot":null,"render_mode":"css","right":null,"right_units":"screen","tags":[],"top":null,"top_units":"screen","x_range_name":"default","y_range_name":"default"},"type":"BoxAnnotation"} but coffee has no such property
CategoricalAxis: defaults are out of sync between Python and CoffeeScript
    CategoricalAxis.ticker: coffee defaults to {"type":"CategoricalTicker"} but python defaults to {"attributes":{"name":null,"tags":[]},"type":"CategoricalTicker"}
    CategoricalAxis.formatter: coffee defaults to {"type":"CategoricalTickFormatter"} but python defaults to {"attributes":{"name":null,"tags":[]},"type":"CategoricalTickFormatter"}
DataRange1d: defaults are out of sync between Python and CoffeeScript
    DataRange1d.start: coffee defaults to 0 but python defaults to null
    DataRange1d.end: coffee defaults to 1 but python defaults to null
DatetimeAxis: defaults are out of sync between Python and CoffeeScript
    DatetimeAxis.ticker: coffee defaults to {"type":"DatetimeTicker"} but python defaults to {"attributes":{"desired_num_ticks":6,"name":null,"num_minor_ticks":0,"tags":[],"tickers":[{"id":"2a94e8e7-9695-43b9-a61c-60d63b8d98f8","type":"AdaptiveTicker"},{"id":"61c6e900-1788-4d94-8389-b43b0820da69","type":"AdaptiveTicker"},{"id":"f55d6b89-0b59-4913-ae71-7ba5da208975","type":"AdaptiveTicker"},{"id":"84b9e331-16f1-4b25-80b9-1fefbaaac552","type":"DaysTicker"},{"id":"ce4cd5bd-7fb8-45f5-9265-e530c203fe33","type":"DaysTicker"},{"id":"c9e3d529-9698-4ed6-af7a-066d862c1cf1","type":"DaysTicker"},{"id":"0a803b58-49f2-43c6-845f-e259f5bfab38","type":"DaysTicker"},{"id":"8acf9c96-2e0d-4138-8d7a-dc0dc11338ed","type":"MonthsTicker"},{"id":"93ddfcd4-7b0d-435c-a70a-e69ad98782cd","type":"MonthsTicker"},{"id":"c2f6e6d4-272e-41ab-9fc8-fefcf3501dad","type":"MonthsTicker"},{"id":"75e2b57f-1dac-4f63-a8a8-9791592d4361","type":"MonthsTicker"},{"id":"e48ee58c-fd14-4c20-86a7-cbac270c9d3c","type":"YearsTicker"}]},"type":"DatetimeTicker"}
    DatetimeAxis.formatter: coffee defaults to {"type":"DatetimeTickFormatter"} but python defaults to {"attributes":{"formats":{},"name":null,"tags":[]},"type":"DatetimeTickFormatter"}
GMapPlot: defaults are out of sync between Python and CoffeeScript
    GMapPlot.tool_events: coffee defaults to {"type":"ToolEvents"} but python defaults to {"attributes":{"geometries":[],"name":null,"tags":[]},"type":"ToolEvents"}
Image: defaults are out of sync between Python and CoffeeScript
    Image.color_mapper: coffee defaults to {} but python defaults to {"attributes":{"high":null,"low":null,"name":null,"palette":["#000000","#252525","#525252","#737373","#969696","#bdbdbd","#d9d9d9","#f0f0f0","#ffffff"],"reserve_color":"#ffffff","reserve_val":null,"tags":[]},"type":"LinearColorMapper"}
LassoSelectTool: coffee is missing some properties found in Python
    LassoSelectTool.overlay: python defaults to {"attributes":{"fill_alpha":{"value":0.5},"fill_color":{"value":"lightgrey"},"level":"overlay","line_alpha":{"value":1},"line_cap":"butt","line_color":{"value":"black"},"line_dash":[4,4],"line_dash_offset":0,"line_join":"miter","line_width":{"value":2},"name":null,"plot":null,"tags":[],"x_range_name":"default","xs":[],"xs_units":"screen","y_range_name":"default","ys":[],"ys_units":"screen"},"type":"PolyAnnotation"} but coffee has no such property
Bokeh: LogTickFormatter not configured with a ticker, using default base of 10 (labels will be incorrect if ticker base is not 10)
LogAxis: defaults are out of sync between Python and CoffeeScript
    LogAxis.ticker: coffee defaults to {"type":"LogTicker"} but python defaults to {"attributes":{"base":10,"desired_num_ticks":6,"mantissas":[1,5],"max_interval":null,"min_interval":0,"name":null,"num_minor_ticks":5,"tags":[]},"type":"LogTicker"}
    LogAxis.formatter: coffee defaults to {"type":"LogTickFormatter"} but python defaults to {"attributes":{"name":null,"tags":[],"ticker":null},"type":"LogTickFormatter"}
PolySelectTool: coffee is missing some properties found in Python
    PolySelectTool.overlay: python defaults to {"attributes":{"fill_alpha":{"value":0.5},"fill_color":{"value":"lightgrey"},"level":"overlay","line_alpha":{"value":1},"line_cap":"butt","line_color":{"value":"black"},"line_dash":[4,4],"line_dash_offset":0,"line_join":"miter","line_width":{"value":2},"name":null,"plot":null,"tags":[],"x_range_name":"default","xs":[],"xs_units":"screen","y_range_name":"default","ys":[],"ys_units":"screen"},"type":"PolyAnnotation"} but coffee has no such property
TableColumn: defaults are out of sync between Python and CoffeeScript
    TableColumn.formatter: coffee defaults to {"type":"StringFormatter"} but python defaults to {"attributes":{"font_style":"normal","name":null,"tags":[],"text_align":"left","text_color":null},"type":"StringFormatter"}
    TableColumn.editor: coffee defaults to {"type":"StringFormatter"} but python defaults to {"attributes":{"completions":[],"name":null,"tags":[]},"type":"StringEditor"}
TileRenderer: defaults are out of sync between Python and CoffeeScript
    TileRenderer.tile_source: coffee defaults to {"type":"WMTSTileSource"} but python defaults to {"attributes":{"attribution":"","extra_url_vars":{},"initial_resolution":156543.03392804097,"max_zoom":30,"min_zoom":0,"name":null,"tags":[],"tile_size":256,"url":"","x_origin_offset":20037508.34,"y_origin_offset":20037508.34},"type":"WMTSTileSource"}
Python/Coffee matching defaults problems: 12
```
Recap, things break down like this:
* ~~~`DataRange1d --- messed up because computed props.~~~  fixed in other PR
* Things like: `coffee defaults to {"type":"ToolEvents"} but python defaults to {"attributes":{"geometries":[],"name":null,"tags":[]},"type":"ToolEvents"}`  I am not sure if this is a real problem  with how the instances are being created on the JS side (e.g., `new ToolEvents.Model()`), or if it just some artifact of the testing process. Advice on how to proceed welcomed. 
* Whatever (if anything) is going on with `Image`

### CODE REVIEW

**Comment by bryevdv** [comment]
@birdsarah 8f1869b seems to fix the geo json example issue, please confirm it is working as expected

File: bokehjs/src/coffee/common/tool_events.coffee [hunk 1]
```diff
@@ -4,5 +4,12 @@ HasProperties = require "./has_properties"
 class ToolEvents extends HasProperties
   type: 'ToolEvents'
 
+  defaults: () ->
+    return {
+      geometries: []
+      name: null
+      tags: []
```
Comment (havocp) [inline bokehjs/src/coffee/common/tool_events.coffee]: why not chain up to get name/tags as most of the other defaults methods do?

Comment (bryevdv) [inline bokehjs/src/coffee/common/tool_events.coffee]: Just an oversight, thanks.

**Comment by havocp** [comment]
```
GMapPlot: defaults are out of sync between Python and CoffeeScript
    GMapPlot.tool_events: coffee defaults to {"type":"ToolEvents"} but python defaults to {"attributes":{"geometries":[],"name":null,"tags":[]},"type":"ToolEvents"}
```

reading code, it looks to me like this is a problem with the tests. In generate_defaults.py we go over the values of the properties and replace each Model with the ref + attributes. In test_defaults.coffee, we do not.

What needs fixing is the lines in test_defaults.coffee that do `attrs = instance.attributes_as_json()` and `strip_ids(attrs)` ... the problem is that `attributes_as_json()` already has only refs, rather than full `HasProperties` objects, so you won't be able to get the attributes of the child objects. I think the fix is probably to get `instance.serializable_attributes()` instead of `attributes_as_json`, and then essentially you'd reimplement `attributes_as_json` in a way that includes the attributes of each instance instead of only the ref... and also omits the ids. If you look in has_properties.coffee, in `@_value_to_json` the line `value.ref()` instead needs to get the ref, delete the id from the ref, then add `attributes` into the ref.

If you want me to address this in a followup PR I can do that

**Comment by bryevdv** [comment]
> If you want me to address this in a followup PR I can do that

Sure that would be helpful, if it i quick/simple for you. What is the change to test turning off sending all the values? I'd like to at least check that out after these are merged.

**Comment by havocp** [comment]
It may be the one line here: https://github.com/bokeh/bokeh/blob/master/bokeh/document.py#L499
In `Document._references_json`

Though possibly that's too simple and there's something that breaks when we flip it to include_defaults=False.

File: bokehjs/src/coffee/common/tool_events.coffee (hunk 1 above)
Comment (bryevdv) [inline bokehjs/src/coffee/common/tool_events.coffee]: changed in #3420

**Pull request merged**